src/setup.rs

```diff
@@ -177,30 +225,79 @@ macro_rules! impl_pins {
     };
 }
 
-#[cfg(feature = "device-selected")]
+#[cfg(all(
+    feature = "device-selected",
+    any(feature = "stm32f4xx-hal", feature = "stm32f7xx-hal")
+))]
 impl_pins!(
     RmiiRefClk: [
-        PA1<Input<Floating>>,
+        PA1<Input>,
     ],
     RmiiCrsDv: [
-        PA7<Input<Floating>>,
+        PA7<Input>,
     ],
     RmiiTxEN: [
-        PB11<Input<Floating>>,
-        PG11<Input<Floating>>,
+        PB11<Input>,
+        PG11<Input>,
     ],
     RmiiTxD0: [
-        PB12<Input<Floating>>,
-        PG13<Input<Floating>>,
+        PB12<Input>,
+        PG13<Input>,
     ],
     RmiiTxD1: [
-        PB13<Input<Floating>>,
-        PG14<Input<Floating>>,
+        PB13<Input>,
+        PG14<Input>,
     ],
     RmiiRxD0: [
-        PC4<Input<Floating>>,
+        PC4<Input>,
     ],
     RmiiRxD1: [
-        PC5<Input<Floating>>,
+        PC5<Input>,
     ],
 );
+
+#[cfg(feature = "stm32f1xx-hal")]
+mod stm32f1 {
+    use super::*;
+    use stm32f1xx_hal::gpio::{gpioa, gpiob, gpioc, gpiod, Alternate, Floating, Input, PushPull};
+
+    // STM32F1xx's require access to the CRL/CRH registers to change pin mode. As a result, we
+    // require that pins are already in the necessary mode before constructing `EthPins` as it
+    // would be inconvenient to pass CRL and CRH through to the `AlternateVeryHighSpeed` callsite.
+
+    type PA1 = gpioa::PA1<Input<Floating>>;
+    type PA2 = gpioa::PA2<Alternate<PushPull>>;
+    type PA7 = gpioa::PA7<Input<Floating>>;
+    type PB11 = gpiob::PB11<Alternate<PushPull>>;
+    type PB12 = gpiob::PB12<Alternate<PushPull>>;
+    type PB13 = gpiob::PB13<Alternate<PushPull>>;
+    type PC1 = gpioc::PC1<Alternate<PushPull>>;
+    type PC4 = gpioc::PC4<Input<Floating>>;
+    type PC5 = gpioc::PC5<Input<Floating>>;
+    type PD8 = gpiod::PD8<Input<Floating>>;
+    type PD9 = gpiod::PD9<Input<Floating>>;
+    type PD10 = gpiod::PD10<Input<Floating>>;
+
+    unsafe impl RmiiRefClk for PA1 {}
+    unsafe impl RmiiCrsDv for PA7 {}
+    unsafe impl RmiiCrsDv for PD8 {}
+    unsafe impl RmiiTxEN for PB11 {}
+    unsafe impl RmiiTxD0 for PB12 {}
+    unsafe impl RmiiTxD1 for PB13 {}
+    unsafe impl RmiiRxD0 for PC4 {}
+    unsafe impl RmiiRxD0 for PD9 {}
+    unsafe impl RmiiRxD1 for PC5 {}
+    unsafe impl RmiiRxD1 for PD10 {}
+
+    macro_rules! impl_alt_very_high_speed {
+        ($($PIN:ident),*) => {
+            $(
+                impl AlternateVeryHighSpeed for $PIN {
+                    fn into_af11_very_high_speed(self) {}
```

Comment:
As I mentioned in the other PR:

It seems that we're relying on the HAL to always set the pin at max speed 50MHz, is that right ? It seems that the HAL indeed does that, but is that a documented behavior ?

@korken89 Table 209 does not guarantee that the pins will be in the right speed, it requires it.

Reply:
Correct, it is made by the mode setting when making a pin `Output`, see here: https://github.com/stm32-rs/stm32f1xx-hal/blob/master/src/gpio.rs#L967

- `0b11` is output @ 50 Mhz
- `0b01` is output @ `0 Mhz
- `0b10` is output @ 2 Mhz

Reply:
Yes, but we should ask the maintainers of the F1 HAL to document this default behavior so we can rely on it, but we don't have to wait for it to merge this PR.

Either way, I think we should document this special requirement for the F1 in the new function.

Reply:
The slew rate is now automatically set to VeryHigh (or MHz50) for `stm32f107` as well